flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/StandaloneCompletedCheckpointStore.java

```diff
@@ -64,9 +62,9 @@ public void recover() throws Exception {
 
 	@Override
 	public void addCheckpoint(CompletedCheckpoint checkpoint) throws Exception {
-		checkpoints.addLast(checkpoint);
+		checkpoints.add(checkpoint);
 		if (checkpoints.size() > maxNumberOfCheckpointsToRetain) {
-			checkpoints.removeFirst().discardState();
+			checkpoints.remove().subsume();
```

Comment:
Manually triggered savepoints for example are not discarded when they are subsumed. The CheckpointProperties constructor is package private (for testing) and only the static creator methods (for persistent checkpoints, regular checkpoints, and manually triggered savepoints) are publicly accessible. Let me add a check to the properties that only allow manual discard if the checkpoint is persisted.